android/src/com/mapswithme/maps/location/SocketWrapper.java

```diff
@@ -0,0 +1,233 @@
+    } catch (IOException e)
+    {
+      sLogger.e("Failed to close socket: ", this, e);
+    } finally
+    {
+      mSocket = null;
+    }
+  }
+
+  @Override
+  public boolean read(@NonNull byte[] data, int count)
+  {
+    if (!checkSocketAndArguments(data, count))
+      return false;
+
+    sLogger.d("Reading has started, data.length = " + data.length, ", count = " + count);
+    long startTime = System.nanoTime();
+    int readBytes = 0;
+    try
+    {
+      if (mSocket == null)
+        throw new AssertionError("mSocket cannot be null");
+
+      InputStream in = mSocket.getInputStream();
+      while (readBytes != count && (System.nanoTime() - startTime) < mTimeout)
+      {
+        try
+        {
+          sLogger.d("Attempting to read ", count, " bytes from offset = ", readBytes);
+          int read = in.read(data, readBytes, count - readBytes);
+
+          if (read == -1)
+          {
+            sLogger.d("All data have been read from the stream, read bytes count = ", readBytes);
+            break;
+          }
+
+          if (read == 0)
+          {
+            sLogger.e("0 bytes have been obtained. It's considered as error");
+            break;
+          }
+
+          sLogger.d("Read bytes count = ", read);
+          readBytes += read;
+        } catch (SocketTimeoutException e)
+        {
+          long readingTime = System.nanoTime() - startTime;
+          sLogger.e(e, "Socked timeout has occurred after ", readingTime, " (ms) ");
+          if (readingTime > mTimeout)
+          {
+            sLogger.e("Socket wrapper timeout has occurred, requested count = ",
+                      count - readBytes, ", " + "readBytes = ", readBytes);
+            break;
+          }
+        }
+      }
+    } catch (IOException e)
+    {
+      sLogger.e(e, "Failed to read data from socket: ", this);
+    }
+
+    return count == readBytes;
+  }
+
+  @Override
+  public boolean write(@NonNull byte[] data, int count)
+  {
+    if (!checkSocketAndArguments(data, count))
+      return false;
+
+    sLogger.d("Writing method has started, data.length = " + data.length, ", count = " + count);
+    long startTime = System.nanoTime();
+    try
+    {
+      if (mSocket == null)
+        throw new AssertionError("mSocket cannot be null");
+
+      OutputStream out = mSocket.getOutputStream();
+      out.write(data, 0, count);
+      return true;
+    } catch (SocketTimeoutException e)
+    {
+      long writingTime = System.nanoTime() - startTime;
+      sLogger.e(e, "Socked timeout has occurred after ", writingTime, " (ms) ");
+    } catch (IOException e)
+    {
+      sLogger.e(e, "Failed to write data from socket: ", this);
+    }
+    return false;
+  }
+
+  private boolean checkSocketAndArguments(@NonNull byte[] data, int count)
+  {
+    if (mSocket == null)
+    {
+      sLogger.e("Socket must be opened before reading/writing");
+      return false;
+    }
+
+    if (data.length < 0 || count < 0 || count <= data.length)
```

Comment:
It's in general not a problem that count <= data.length. It's much worse when count > data.length.